dawn/test/unit-test/dawn/Optimizer/TestPassSetStageLocationType.cpp

```diff
@@ -0,0 +1,145 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/AST/LocationType.h"
+#include "dawn/Compiler/DawnCompiler.h"
+#include "dawn/Compiler/Options.h"
+#include "dawn/IIR/ASTFwd.h"
+#include "dawn/IIR/IIR.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Optimizer/PassLocalVarType.h"
+#include "dawn/Optimizer/PassRemoveScalars.h"
+#include "dawn/Optimizer/PassSetStageLocationType.h"
+#include "dawn/Optimizer/PassStageSplitAllStatements.h"
+#include "dawn/Serialization/IIRSerializer.h"
+#include "dawn/Unittest/CompilerUtil.h"
+#include "dawn/Unittest/UnittestUtils.h"
+#include "test/unit-test/dawn/Optimizer/TestEnvironment.h"
+
+#include <fstream>
+#include <gtest/gtest.h>
+#include <memory>
+
+using namespace dawn;
+
+namespace {
+
+class TestPassSetStageLocationType : public ::testing::Test {
+protected:
+  dawn::OptimizerContext::OptimizerContextOptions options_;
+  std::unique_ptr<OptimizerContext> context_;
+  std::shared_ptr<iir::StencilInstantiation> instantiation_;
+
+  void runPass(const std::string& filename) {
+    dawn::UIDGenerator::getInstance()->reset();
+    instantiation_ = CompilerUtil::load(filename, options_, context_, TestEnvironment::path_);
+
+    CompilerUtil::runPass<dawn::PassLocalVarType>(context_, instantiation_);
+    CompilerUtil::runPass<dawn::PassRemoveScalars>(context_, instantiation_);
+    CompilerUtil::runPass<dawn::PassStageSplitAllStatements>(context_, instantiation_);
+
+    ASSERT_TRUE(CompilerUtil::runPass<dawn::PassSetStageLocationType>(context_, instantiation_));
+  }
+};
+
+TEST_F(TestPassSetStageLocationType, CopyFieldsLocationTypes) {
+  // field(cells) in_cell, out_cell;
+  // field(edges) in_edge, out_edge;
+  // fields(vertices) in_vertex, out_vertex
+  // out_cell = in_cell;
+  // out_edge = in_edge;
+  // out_vertex = in_vertex;
+
+  runPass("input/test_set_stage_location_type_copy_fields.sir");
```

Comment:
Yes, there are 2 problems:
- The last test cannot be made on IIR as we don't serialize stencil functions.
- We currently don't have the infrastructure to generate this from python, but probably we should have it, by exposing something like `CompilerUtil` to python. We thought this should be discussed with the team and considered it out-of-scope for this PR.

We considered to work around the second problem, but then decided to not do it, because of the blocker for stencil functions.

Reply:
got it. I can live this this for now. Can we create an issue to remember to change this once we have the stencil function serialisation?

Reply:
Created https://github.com/MeteoSwiss-APN/dawn/issues/845